### PR TITLE
[iOS] Fixed #6082 

### LINKED ISSUE #6082: [Rendering] [iOS] ACR button rendering Is wonky as the SDK sets up height constraint and so does the Teams app
# Target Platform

Please list the platform(s) that your application is targeting and remove platforms you are not targeting:
- iOS

# Target Application
AC iOS SDK v 1.5, pod 2.8.0-beata1

Please provide the name of the application that is rendering the card (ex: Outlook, Teams, etc)
Teams

# Problem Description
When upgrading from 2.4.6 AC SDK to 2.8.0-beata1. It looks like the ACR button being used on Teams SDK gets rendered with the text cutoff, and also the button returns a wrong "Intrinsic Size", which is what the Teams SDK uses to give the button a width constraint.

Mobile Platform Bellevue is working on upgrading the current AC SDK on Teams to v1.5 which is branch release/21.04 and the pod version is 2.8.0-beata1.

There are a couple things

1. We noticed that there have been changes in the ACRButton adds a height constraint to the ACRButton

[button.heightAnchor constraintGreaterThanOrEqualToAnchor:button.titleLabel.heightAnchor constant:button.contentEdgeInsets.top + button.contentEdgeInsets.bottom].active = YES;

And every custom renderer we have on the Teams app, uses ACRButton and then adds our own heigh Constraint, now these two constrains clash and make the button appear weirdly(Screenshots attached)

2. We also noticed that the ACRButtonExpandable.xib and ACRButton.xib have been changed recently and the Teams app actually uses "button.titleLabel.IntrinsicSize" to assign a width constrain to the button for every customrender and the change made in the AC SDK for these xibs(Or the new constraints) return the intrinsicSize to be wrong. So this ends up making the button look wonky.

**The intrinsicSize issue could be potentially fixed on the Teams app relying on the titleLabel.text which would fix the buttons to show the entire text horizontally.

But the vertical part, since the constraints are clashing if it were to be fixed on Teams app, then we'd need to employ a hacky way of removing Height constraints on the ACRButton.**

Please enter a description of the issue. If you just have a question, please post [on StackOverflow instead](https://stackoverflow.com/questions/tagged/adaptive-cards)

# Screenshot

Please provide a screenshot of the problem (if possible)
Screen Shot 2021-07-06 at 11 02 22 AM

# Card JSON

{"type":"AdaptiveCard","version":"1.0","body":[{"type":"RichTextBlock","inlines":[{"type":"TextRun","text":"Please select your country code from the drop down and enter your number in the field below (no special characters allowed) ","wrap":true,"size":"Small","color":"Accent"},{"type":"TextRun","text":" *","color":"Attention","weight":"Bolder"}],"isVisible":true},{"type":"ColumnSet","columns":[{"type":"Column","width":"stretch","items":[{"type":"Input.ChoiceSet","placeholder":"Pick a country code","choices":[{"title":"Choice 1","value":"Choice 1"},{"title":"Choice 2","value":"Choice 2"}],"id":"country"}]},{"type":"Column","width":"stretch","items":[{"type":"Input.Text","placeholder":"Type the number here","id":"NewPhoneNumber"}]}],"isVisible":true},{"type":"RichTextBlock","inlines":[{"type":"TextRun","text":"Set privacy ","wrap":true,"size":"Small","color":"Accent"},{"type":"TextRun","text":" *","color":"Attention","weight":"Bolder"}],"isVisible":true},{"type":"Input.ChoiceSet","placeholder":"Placeholder text","choices":[{"title":"This is  json sample for the Test)","value":"isPrimary"},{"title":"This is  json sample for the Test","value":"ispublic"}],"style":"expanded","wrap":true,"spacing":"Small","id":"Privacy","value":"","isVisible":true},{"type":"ActionSet","actions":[{"type":"Action.Submit","title":"Submit","data":{"Action":"Test"}},{"type":"Action.Submit","title":"Cancel","data":{"Action":"cancel"}}],"isVisible":true},{"type":"TextBlock","text":"Please refer to the following rules and enter the correct number above.","wrap":true,"color":"Attention","isVisible":false},{"type":"FactSet","facts":[{"title":"Allowed no. of digits","value":"{AllowedDigit}"}],"isVisible":false,"separator":true},{"type":"TextBlock","text":"No special characters allowed.","isVisible":false,"weight":"Bolder"},{"type":"TextBlock","text":"Please ﬁll all mandatory ﬁelds above and 'Submit' again.\n","wrap":true,"color":"Attention","isVisible":false},{"type":"TextBlock","text":"Please enter the correct format above and try again.","wrap":true,"color":"Attention","isVisible":false}],"$schema":"http://adaptivecards.io/schemas/adaptive-card.json"}

# Sample Code

If you add 
`[button.heightAnchor constraintEqualToConstant:TSkButtonHeight].active = YES;`

after line number 36 in ACRActionShowCardRenderer.mm file you should see the issue as well
# Version of SDK

What version are you using? Ex: NuGet 1.0.2, or latest main, etc...
`release/21.04` branch and pod version 2.8.0-beata1

### PR DESCRIPTION
**Pull request by jwoo-msft**
# Related Issue

Fixed #6082 

# Description

1. Added button height and title width constraints to public interface to help custom styling.
2. Added mocks to aid in current & future unit tests.

# Sample Card

N/A

# How Verified

How you verified the fix, including one or all of the following:
1. New unit tests were added.
2. All unit tests were run.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/6158)

### CODE REVIEW

**Review by paulcam206 — APPROVED** [review]

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewPrivate.h [hunk 1]
```diff
@@ -77,4 +77,7 @@ typedef void (^ObserverActionBlockForBaseAction)(NSObject<ACOIResourceResolver>
 - (ACOInputResults *)dispatchAndValidateInput:(ACRColumnView *)parent;
 
 - (void)removeObserver:(NSObject *)observer forKeyPath:(NSString *)path onObject:(NSObject *)object;
+
+- (void)setContext:(ACORenderContext *)context;
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewPrivate.h:81]: >(void)setContext:(ACORenderContext *)context; [](http://example.com/codeflow?start=2&length=45)

what is the purpose of this Context setter? From the mock sample I get the usage could cover scenarios as knowing all actions in an action set having icons but I don't see the actual method being enforced in any test

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCardsTests/Mocks/MockACRView.mm [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+@implementation MockACRView
+
+- (instancetype)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
+    return self;
+}
+
+- (void)setMockContext:(MockContext *)context
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCardsTests/Mocks/MockACRView.mm:20]: >setMockContext [](http://example.com/codeflow?start=8&length=14)

This method is never called

Comment (jwoo-msft) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCardsTests/Mocks/MockACRView.mm:20]: it's objective-c syntax. it's get called via @property

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewPrivate.h (hunk 1 above)
Comment (jwoo-msft) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewPrivate.h:81]: context is proving the value for icon placement. the mock value will allow choosing the execution branches.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCardsTests/AdaptiveCardsActionsTest.mm [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+- (void)testACRButtonPublicConstraintsWithImageViewWithKey
+{
+    std::string iconURL = "https://adaptivecards.io/content/cats/1.png";
+    NSString *iconURLObjc = [NSString stringWithCString:iconURL.c_str() encoding:NSUTF8StringEncoding];
+
+    UIImage *image = [UIImage imageWithData:[NSData dataWithContentsOfURL:[NSURL URLWithString:iconURLObjc]]];
+    UIImageView *imageView = [[UIImageView alloc] initWithImage:image];
+    ACOBaseActionElement *acoAction = [self buildSimpleAction];
+    acoAction.element->SetIconUrl(iconURL);
+    NSNumber *number = [NSNumber numberWithUnsignedLongLong:(unsigned long long)acoAction.element.get()];
```
Comment (almedina-ms) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCardsTests/AdaptiveCardsActionsTest.mm:86]: >get [](http://example.com/codeflow?start=98&length=3)

what does this get() return? is it a pointer to the object?

**Review by almedina-ms — APPROVED** [review]

**Pull request merged**